src/rgw/rgw_putobj_processor.cc

```diff
@@ -497,13 +497,23 @@ int MultipartObjectProcessor::complete(size_t accounted_size,
     return r;
   }
 
-  encode(info, bl);
-
   std::unique_ptr<rgw::sal::Object> meta_obj =
     head_obj->get_bucket()->get_object(rgw_obj_key(mp.get_meta(), std::string(), RGW_OBJ_NS_MULTIPART));
   meta_obj->set_in_extra_data(true);
 
-  r = meta_obj->omap_set_val_by_key(dpp, p, bl, true, null_yield);
+  rgw_raw_obj meta_raw_obj;
+  dynamic_cast<rgw::sal::RadosObject*>(meta_obj.get())->get_raw_obj(&meta_raw_obj);
```

Comment:
this line doesn't make sense? 1) seems you want to assign the downcast pointer to a variable?  this must warn?  also, 2) using dynamic_cast implies that the downcast can fail, if the returned object pointer is of a different descendant type--in that case, the returned pointer will be null, so we don't want to operate on it

Reply:
oh, I see.  however, I think the downcast is still an issue (maybe in both places, I don't know)

Reply:
However, a lingering concern is that MultipartObjectProcessor::complete() seems to try to be generic and not Rados specific. However, the first few lines of this function already directly refer to RGWRados. I suspect that it isn't as generic as hoped.

Reply:
I think there should at least be error checking from the dynamic_cast

Reply:
clearly, I agree with you, but it *will* be fully generic in due course

Reply:
you could leave a comment here or at the other location, maybe, to highlight the issue

Reply:
Instead of using dynamic_cast, I changed to call obj_to_raw() in the 2nd commit. The whole purpose is to access more information from the raw obj.